src/core/EdFi.DataManagementService.Core/Middleware/DisallowDuplicateReferencesMiddleware.cs

```diff
@@ -0,0 +1,145 @@
+                StatusCode: 400,
+                JsonSerializer.Serialize(
+                    ForDataValidation(
+                        "Data validation failed. See 'validationErrors' for details.",
+                        traceId: context.FrontendRequest.TraceId,
+                        validationErrors,
+                        []
+            ),
+                    SerializerOptions
+                ),
+                Headers: []
+            );
+            return;
+        }
+
+        await next();
+    }
+
+    private static void ValidateDuplicates<T>(
+        IEnumerable<T> items,
+        Func<T, Guid> getReferentialId,
+        Func<T, string> getPath,
+        Dictionary<string, string[]> validationErrors,
+        int initialPosition
+    )
+    {
+        var seenItems = new HashSet<Guid>();
+        int position = initialPosition;
+
+        foreach (var item in items)
+        {
+            Guid referentialId = getReferentialId(item);
+
+            if (seenItems.Contains(referentialId))
+            {
+                string path = getPath(item);
+
+                // the propertyName varies according to the origin (DescriptorReference or DocumentReferences)
+                string propertyName = path.StartsWith("$", StringComparison.InvariantCultureIgnoreCase)
+                    ? path
+                    : $"$.{path}";
+
+                path = path.StartsWith("$", StringComparison.InvariantCultureIgnoreCase)
+                    ? ExtractArrayName(path)
+                    : path;
+
+                string errorMessage =
+                    $"The {GetOrdinal(position)} item of the {path} has the same identifying values as another item earlier in the list.";
+
+                if (validationErrors.ContainsKey(propertyName))
+                {
+                    var existingMessages = validationErrors[propertyName].ToList();
+                    existingMessages.Add(errorMessage);
+                    validationErrors[propertyName] = existingMessages.ToArray();
+                }
+                else
+                {
+                    validationErrors[propertyName] = new[] { errorMessage };
+                }
+            }
+            else
+            {
+                seenItems.Add(referentialId);
+            }
+            position++;
+        }
+    }
+
+    private static string GetOrdinal(int number)
+    {
+        if (number % 100 == 11 || number % 100 == 12 || number % 100 == 13)
+            return $"{number}th";
+
+        return (number % 10) switch
+        {
+            1 => $"{number}st",
+            2 => $"{number}nd",
+            3 => $"{number}rd",
+            _ => $"{number}th"
+        };
+    }
+
+    private static string ExtractArrayName(string path)
+    {
+        // Logic to extract the array name from the JSON path, e.g., "gradeLevels".
+        var parts = path.Split('.');
+        return parts.Length > 1 ? parts[1].Trim('[', ']', '*') : string.Empty;
+    }
+}
```

Comment:
@JBrenesSimpat 
Could you please add unit tests for this newly added middleware ?